cocotb/share/lib/fli/FliImpl.cpp

```diff
@@ -42,13 +42,10 @@ static FliImpl         *fli_table;
 
 void FliImpl::sim_end()
 {
-    if (GPI_DELETE != sim_finish_cb->get_call_state()) {
```

Comment:
Whose job is it to deregister the callback now?

Reply:
> This is necessary if there are multiple GPI users. If we deregister the shutdown callback, not every GPI user would be notified of a shutdown

Can you elaborate on this?

Reply:
Previously, when `sim_end` was called, it deregistered the shutdown callback. I assume this was done to prevent shutdown behavior from running *twice* because it currently does shutdown behavior immediately (after the stack unravels). This isn't a problem if we put all shutdown behavior in the shutdown callback and don't deregister it.

The reason this is necessary for multiple GPI users is in the case of one user calling for shutdown, it must notify all the other users of shutdown using `gpi_sim_event`. If we call this when `sim_end` is called, we get re-entrance. Placing this in the shutdown callback means it's always called from a simulator context.